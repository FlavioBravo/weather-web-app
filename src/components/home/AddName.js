import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddName = ({ setCities }) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            // setCities(cities => [inputValue, ...cities]);
            setCities(cities => [inputValue]);
            setInputValue('');
        }

    }

    return (
        <form onSubmit={handleSubmit}>
            <h4>Filters</h4>
            <label>Name:</label>
            <p> {inputValue} </p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

AddName.propTypes = {
    setCities: PropTypes.func.isRequired
}