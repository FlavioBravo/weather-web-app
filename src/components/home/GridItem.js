import React from 'react';
import PropTypes from 'prop-types';
import { useHistory } from "react-router-dom";

export const GridItem = ({ id, name, state, country, coord }) => {

    const history = useHistory();

    const handleForecast = () => {
        history.push(`/forecast/${id}`);
    };

    return (
        <tr className="animate__animated animate__fadeIn" onClick={ handleForecast } >
            <td> { id } </td>
            <td> { name } </td>
            <td> { state } </td>
            <td> { country } </td>
            <td> { coord.lon } </td>
            <td> { coord.lat } </td>
        </tr>
    )
}

GridItem.propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    state: PropTypes.string.isRequired,
    country: PropTypes.string.isRequired,
    coord: PropTypes.object.isRequired,
}