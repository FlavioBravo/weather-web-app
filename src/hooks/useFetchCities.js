import { useState, useEffect } from 'react'
import { getCities } from '../helpers/getCities';


export const useFetchCities = ( city ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {

        getCities( city )
            .then( imgs => {
                
                setState({
                    data: imgs,
                    loading: false
                });
            })

    }, [city])




    return state;


}