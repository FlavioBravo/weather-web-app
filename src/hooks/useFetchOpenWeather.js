import { useState, useEffect } from 'react';
import { get5daysForecast } from '../helpers/get5daysForecast';


export const useFetchOpenWeather = ( cityId ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {
        let isSubscribed = true;
        get5daysForecast( cityId )
            .then( forecast => {
                if (isSubscribed) {
                    setState({
                        data: forecast,
                        loading: false
                    });
                }
            })
        return () => isSubscribed = false;
    }, [cityId])

    return state;

}