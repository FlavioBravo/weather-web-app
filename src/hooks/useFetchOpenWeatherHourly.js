import { useState, useEffect } from 'react';
import { getHourlyForecast } from '../helpers/getHourlyForecast';


export const useFetchOpenWeatherHourly = ( lat, lon ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {
        let isSubscribed = true;
        getHourlyForecast( lat, lon )
            .then( hourly => {
                if (isSubscribed) {
                    setState({
                        data: hourly,
                        loading: false
                    });
                }
            })
        return () => isSubscribed = false;
    }, [lat, lon])

    return state;

}