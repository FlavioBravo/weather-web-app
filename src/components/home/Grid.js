import React from 'react';
import PropTypes from 'prop-types';

import { useFetchCities } from '../../hooks/useFetchCities';
import { GridItem } from './GridItem';

export const Grid = ({ city }) => {

    const { data: images, loading } = useFetchCities(city);

    return (
        <>
            <br />
            <h3 className="animate__animated animate__fadeIn"> {city} </h3>
            <br />
            { loading && <p className="animate__animated animate__flash">Loading</p>}

            <table className="">
                <thead>
                    <tr>
                        <th>ID</th>
                        <th>Name</th>
                        <th>State</th>
                        <th>Country</th>
                        <th>Longitude</th>
                        <th>Latitude</th>
                    </tr>
                </thead>
                <tbody>
                    {
                        images.map(img => (
                            <GridItem
                                key={img.id}
                                {...img}
                            />
                        ))
                    }
                </tbody>
            </table>
        </>
    )
}

Grid.propTypes = {
    city: PropTypes.string.isRequired
}