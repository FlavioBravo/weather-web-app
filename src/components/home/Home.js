import React, { useState } from 'react';

import { AddName } from './AddName';
import { Grid } from './Grid';

export const Home = ({ defaultCities = [] }) => {

    const [cities, setCities] = useState( defaultCities );

    return (
      <>
          <h2>Weather Web App</h2>
          <AddName setCities={ setCities } />
          <hr />

          <ol>
              {
                  cities.map( city  => (
                      <Grid 
                          key={ city }
                          city={ city }
                      />
                  ))
              }
          </ol>

      </>
  )
}
