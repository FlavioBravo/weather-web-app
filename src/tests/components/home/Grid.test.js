import React from 'react';
import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import { Grid } from '../../../components/home/Grid';
import { useFetchCities } from '../../../hooks/useFetchCities';
jest.mock('../../../hooks/useFetchCities');


describe('Pruebas en el <Grid />', () => {

    const city = 'Tum';


    test('debe de mostrarse correctamente', () => {

        useFetchCities.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow( <Grid city={ city } /> );
        expect( wrapper ).toMatchSnapshot();
        
    });

    test('debe de mostrar items cuando se cargan de useFetchCities', () => {
        
        const cities = [{
            id: 3691146,
            name: "Tumbes",
            state: "",
            country: "PE",
            coord: {
                lon: -80.5,
                lat: -3.83333
            }
        },
        {
            id: 3691148,
            name: "Tumbes",
            state: "",
            country: "PE",
            coord: {
                "lon": -80.441391,
                "lat": -3.56667
            }
        },
        {
            id: 3691150,
            name: "Tum√°n",
            state: "",
            country: "PE",
            coord: {
                "lon": -79.700562,
                "lat": -6.74778
            }
        }];

        useFetchCities.mockReturnValue({
            data: cities,
            loading: false
        });

        const wrapper = shallow( <Grid city={ city } /> );
        
        // expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GridItem').length ).toBe( cities.length );
    })
    
    

})