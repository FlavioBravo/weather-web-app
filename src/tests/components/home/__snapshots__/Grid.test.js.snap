// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <Grid /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <br />
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Tum
     
  </h3>
  <br />
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <table
    className=""
  >
    <thead>
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          State
        </th>
        <th>
          Country
        </th>
        <th>
          Longitude
        </th>
        <th>
          Latitude
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</Fragment>
`;
